ingest-router/src/relay_project_config_handler.rs

```diff
@@ -0,0 +1,1013 @@
+struct RelayProjectConfigsRequest {
+    /// DSN public keys to fetch configs for.
+    #[serde(rename = "publicKeys")]
+    pub public_keys: Vec<String>,
+
+    /// Whether to include global config (optional).
+    ///
+    /// first upstream gets original value, others get `Some(false)`.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub global: Option<bool>,
+
+    /// Other fields (`noCache`, future fields) for forward compatibility.
+    #[serde(flatten)]
+    pub extra: HashMap<String, JsonValue>,
+}
+
+impl RelayProjectConfigsRequest {
+    fn from_bytes(bytes: &Bytes) -> Result<Self, serde_json::Error> {
+        serde_json::from_slice(bytes)
+    }
+
+    fn to_bytes(&self) -> Result<Bytes, serde_json::Error> {
+        let json = serde_json::to_vec(self)?;
+        Ok(Bytes::from(json))
+    }
+}
+
+/// Response format for the relay project configs endpoint.
+///
+/// See module-level docs for merge strategy and field details.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct RelayProjectConfigsResponse {
+    /// Project configs (HashMap merged from all upstreams).
+    pub configs: HashMap<String, JsonValue>,
+
+    /// Keys being computed async or from failed upstreams (concatenated).
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub pending: Option<Vec<String>>,
+
+    /// Other fields (`global`, `global_status`, future fields).
+    #[serde(flatten)]
+    pub extra: HashMap<String, JsonValue>,
+}
+
+impl RelayProjectConfigsResponse {
+    fn from_bytes(bytes: &Bytes) -> Result<Self, serde_json::Error> {
+        serde_json::from_slice(bytes)
+    }
+}
+
+/// Result from an upstream task request.
+struct UpstreamTaskResult {
+    /// The upstream URL that was contacted.
+    sentry_url: String,
+    /// The public keys that were requested from this upstream.
+    public_keys: Vec<String>,
+    /// The result of the upstream request.
+    result: Result<Response<Bytes>, IngestRouterError>,
+}
+
+/// Merged results from all upstream tasks.
+struct MergedResults {
+    /// All configs merged from successful upstreams.
+    configs: HashMap<String, JsonValue>,
+    /// All pending keys (from failed upstreams or upstream pending arrays).
+    pending: Vec<String>,
+    /// Extra fields (global config, status, etc.).
+    extra: HashMap<String, JsonValue>,
+}
+
+/// Task handle paired with its public keys for graceful failure handling.
+///
+/// The public keys are tracked outside the task so they can be added to the
+/// pending array if the task times out or panics, maintaining v3 protocol compliance.
+struct TaskWithKeys {
+    /// The spawned task handle.
+    handle: JoinHandle<UpstreamTaskResult>,
+    /// Public keys requested from this upstream.
+    public_keys: Vec<String>,
+}
+
+/// Handler for Relay Project Configs endpoint.
+///
+/// See module-level docs for complete protocol details, implementation strategy,
+/// and request/response flow diagrams.
+pub struct RelayProjectConfigsHandler {
```

Comment:
Since there will be a number of handlers that follow this pattern could this be a trait to enforce they all follow similar patterns?

e.g. something like
```
trait Handler<Req: Serialize + Deserialize, Res: Serialize + Deserialize> {
    fn split_requests(request: Req) -> Vec<Req>,
    fn merge_results(results: Vec<Res>) -> Res,
}
```

where Req and Res can be anything that implement serialize+deserialize. I think this forces the implementation of each endpoint to conform to some basic patterns and look the same which is nice for readability. It could enable more of the the common parts - spawning multiple requests and waiting for them, timeouts, error handling/backpressure to be done in a central place outside of the specific handler.

Reply:
I do not know of more use cases where this may be needed right now. If we see that use case pop up, then I can make this a generic trait and have multiple implementations. But I would rather do it when I see the need for it.